rfd/0124-plugin-secrets.md

```diff
@@ -0,0 +1,174 @@
+
+## Why
+
+As we grow our plugin support, we have discovered that the various services we
+hope to integrate do not offer a uniform mechanism for configuration. For instance,
+Slack uses oauth2, but Okta requires an API token, and Jamf requires a username and
+password.
+
+Today, the plugin houses a credentials object that allows for this sort of storage.
+However, unlike an oauth authorization code, static credentials pose a much greater risk of
+exposure, and rotating credentials can be a time consuming process.  To reduce the risks associated
+with storing static credentials within Teleport and to allow for easier credentials rotation,
+we would like to create a new resource within Teleport that is handled in an explicitly secure
+way.
+
+## Details
+
+### The `PluginStaticCredentials` object
+
+A new object will be needed to house static credentials. At present it will support:
+
+- API tokens
+- Usernames and passwords
+- OAuth client ID and secret
+
+This object has very strict access control requirements that will be explained in detail
+in the **Access Control** section. On creation, the plugin and static credentials will
+receive a UUID called the `plugin-label`. The static credentials will have an internal
+label that receives this value. Additionally, one plugin may have _many_ credentials
+that are disambiguated with an additional label `plugin-credentials-type`. This is an
+arbitrary plugin specific string that will allow plugin authors to refer to multiple
+static
+credentials if needed.
+
+```mermaid
+flowchart LR
+  Plugin
+  Okta[Okta Credentials]
+  Slack[Slack Credentials]
+  OpsGenie[OpsGenie Credentials]
+
+  Plugin-->|credentials type okta|Okta
+  Plugin-->|credentials type slack|Slack
+  Plugin-->|credentials type opsgenie|OpsGenie
+```
+
+```yaml
+kind: plugin_static_credentials
+version: v1
+metadata:
+  name: credentials-name
+  labels:
+    teleport.internal/plugin-label: 48398071-9860-4be6-9f29-ca6ca5bc7155 # unique plugin label.
+    teleport.internal/plugin-credentials-type: slack # a unique type associated with the credentials
+spec:
+  credentials:
+    # Only one of these credential types must be defined at a time. All values in this
+    # are string literals and not file locations.
+    api_token: example-token
+    basic_auth:
+      username: example-user
+      password: example-password
+    oauth_client_secret:
+      client_id: example-client-id
+      client_secret: example-client-secret
+```
+
+If a `Plugin` is using oauth, however, it will not need any `PluginStaticCredentials` objects.
+
+### Changes to the `Plugin` object
+
+The `Plugin` object's `Credentials` field will have a new type of credential called
+`PluginStaticCredentialsRefs` that contains a UUID and a list of plugin types.
+
+```yaml
+plugin_static_credentials_refs:
+  label: 48398071-9860-4be6-9f29-ca6ca5bc7155
+  types:
+    - okta
+    - slack
+    - opsgenie
```

Comment:
I think `types` are redundant and a bit confusing tbh. I'm not sure I understand what purpose they serve, isn't matching by labels enough?

Reply:
You touched on this in your later comment, I was trying to be too specific with the label usage.